### ISSUE TITLE
Intermittent UI Freezing on Startup

### ISSUE THREAD

**Issue by tsc-awardle** [mdsecresearch/BurpSuiteSharpener#89]
Hello again!

Not sure if I'm the only one with this problem, but my Burp window occasionally freezes on startup if the Sharpener extension is loaded. Burp needs to be manually killed before it works again.

It only happens around ~50% of the time, but I've noticed it only occurs if Sharpener is loading at the same time as other heavy extensions (most notably Burp Bounty Pro and Logger++). Once Burp is loaded, everything works as expected and it never freezes, even while manually reloading extensions one at a time.

During a freeze, Burp infinitely prints the following exception to the console:
```
java.lang.NullPointerException: Cannot read field "preferred" because "this.xTotal" is null
        at java.desktop/javax.swing.BoxLayout.preferredLayoutSize(BoxLayout.java:294)
        at java.desktop/java.awt.Container.preferredSize(Container.java:1826)
        at java.desktop/java.awt.Container.getPreferredSize(Container.java:1810)
        at java.desktop/javax.swing.JComponent.getPreferredSize(JComponent.java:1727)
        at java.desktop/java.awt.BorderLayout.preferredLayoutSize(BorderLayout.java:729)
        at java.desktop/java.awt.Container.preferredSize(Container.java:1826)
        at java.desktop/java.awt.Container.getPreferredSize(Container.java:1810)
        at java.desktop/javax.swing.JComponent.getPreferredSize(JComponent.java:1727)
        at java.desktop/javax.swing.ScrollPaneLayout.layoutContainer(ScrollPaneLayout.java:799)
        at java.desktop/java.awt.Container.layout(Container.java:1541)
        at java.desktop/java.awt.Container.doLayout(Container.java:1530)
        at java.desktop/java.awt.Container.validateTree(Container.java:1725)
        at java.desktop/java.awt.Container.validateTree(Container.java:1734)
        at java.desktop/java.awt.Container.validateTree(Container.java:1734)
        at java.desktop/java.awt.Container.validate(Container.java:1660)
        at java.desktop/javax.swing.plaf.basic.BasicTabbedPaneUI.ensureCurrentLayout(BasicTabbedPaneUI.java:1651)
        at java.desktop/javax.swing.plaf.basic.BasicTabbedPaneUI.getTabRunCount(BasicTabbedPaneUI.java:1677)
        at com.formdev.flatlaf.ui.FlatTabbedPaneUI.ensureCurrentLayout(FlatTabbedPaneUI.java:1607)
        at com.formdev.flatlaf.ui.FlatTabbedPaneUI.paint(FlatTabbedPaneUI.java:1051)
        at java.desktop/javax.swing.plaf.ComponentUI.update(ComponentUI.java:161)
        at com.formdev.flatlaf.ui.FlatTabbedPaneUI.update(FlatTabbedPaneUI.java:1040)
        at java.desktop/javax.swing.JComponent.paintComponent(JComponent.java:842)
        at java.desktop/javax.swing.JComponent.paint(JComponent.java:1119)
        at java.desktop/javax.swing.JComponent.paintToOffscreen(JComponent.java:5311)
        at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBufferedImpl(RepaintManager.java:1657)
        at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBuffered(RepaintManager.java:1632)
        at java.desktop/javax.swing.RepaintManager$PaintManager.paint(RepaintManager.java:1570)
        at java.desktop/javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:262)
        at java.desktop/javax.swing.RepaintManager.paint(RepaintManager.java:1337)
        at java.desktop/javax.swing.JComponent._paintImmediately(JComponent.java:5259)
        at java.desktop/javax.swing.JComponent.paintImmediately(JComponent.java:5069)
        at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:879)
        at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:862)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:862)
        at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:835)
        at java.desktop/javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:784)
        at java.desktop/javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1898)
        at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:771)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:741)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

After some trial and error, I believe it's caused by [BurpFrameListeners](https://github.com/mdsecresearch/BurpSuiteSharpener/blob/main/src/main/java/com/irsdl/burp/sharpener/actitivities/ui/burpFrame/BurpFrameListeners.java). I commented out all references to that class and haven't had a freeze since.

I have occasionally seen the popups from this feature telling me the window is slightly off-screen, and coincidentally this is almost the exact same time that the window freezes. Is there any reason to keep the feature, or can it be removed from the code? If others are relying on it, then maybe it's best to give an option to disable it in the settings.

**Comment by irsdl** [comment]
Thanks for bringing this up. Could you please untick those two options in the screenshot below and tell me whether that will avoid the affected code:
![image](https://user-images.githubusercontent.com/1798313/205464525-dc2c877c-1fcf-47ba-af9c-ee047f094220.png)

The error happens probably as a result of long delay in loading the extension UI so it cannot get access to the top menu. I have a feeling that this should even prevent the extension to add itself to the top menu which will be worse as it means it needs to be unloaded and loaded later when things are settled down.
My plan at the moment is to migrate to use Montoya API completely but as it is a big change, I may not be able to release short fixes in between. Whilst Burp says it is compatible with both at the same time for a smooth upgrade, it is still not possible to use the same function we had previously for the new API and it needs to be coded completely for the new version which means we will have two functions for each action if we still want to keep the old version which is not ideal for debugging and code fixes in the future.

**Comment by tsc-awardle** [comment]
Interesting, I will test that and report back.

Meanwhile, I did encounter freezing issues again, but they're different this time. Yesterday, burp locked up again on startup, and I have a hunch it was caused by the [top menu un-bolding timer](https://github.com/mdsecresearch/BurpSuiteSharpener/blob/main/src/main/java/com/irsdl/burp/sharpener/actitivities/ui/topMenu/TopMenu.java#L408-L428). I commented out that code and haven't had another freeze on startup yet.

I was also able to reproduce a separate freezing issue with the following steps:
- Start burp with a few heavy extensions loaded (Sharpener, Logger++, Burp Bounty Pro)
- Unload Logger++
- Load Logger++
- Entire burp UI freezes and becomes unusable

It's interesting though, because it never occurs when **reloading** the extension (CTRL+click). It only occurs when fully unloading, and then loading the extension again (click+click).

It's very hard to isolate the issue, since the stack trace doesn't point to any particular extension. The only hint is that the issue occurs in an calls to the event dispatch thread (e.g., `SwingUtilities.invokeLater`), and that it's trying to access things that don't exist (almost always an NPE in my stack traces). I tend to suspect async calls in these situations, as there's no guarantee the GUI is in the same state in future calls.

Would be awesome if PortSwigger could gave us a proper API interface to interact with other GUI elements like top JMenu.

Either way, I will mark this PR as WIP and continue to work on it.

**Comment by irsdl** [comment]
Haha, thanks. I think the problem we have here is that Sharpener cannot get hold of the UI which is needed to be traversed in order to obtain the necessary objects. Also as it adds some listeners to the different elements, if a UI becomes unresponsive, it may go to a unknown criteria (so anything may happen including a total UI freeze by going through an infinite loop). I think in the next version, I should define a time out for all processes and come up with a better way to not run a method multiple times in a short period. I don't think PortSwigger will ever release something for the extensions to control the UI. They have a very large backlog of seriously important stuff to address but even some those are in their list for years.

**Comment by tsc-awardle** [comment]
I've been running sharpener with the [BurpFrameListeners disabled](https://github.com/tsc-awardle/BurpSuiteSharpener/tree/fix-UI-freeze-on-startup) the past couple weeks. I also unticked those options in the settings. I haven't noticed any issues in my testing.

I did encounter another freeze on startup today, but it's definitely happening less often. Has anyone else been able to reproduce the freezing issue, or am I the only one?
